### ISSUE TITLE
Spanner Sessions Not Getting Cleaned Up

### ISSUE THREAD

**Issue by jon-whit** [googleapis/google-cloud-go#1616]
## Client

Spanner Golang

## Describe Your Environment

Alpine Docker on GKE

----

I have some code that looks like this:

```
type SpannerDatastore struct {
	*spanner.Client
}

func (ds *SpannerDatastore) ListEntitlements(ctx context.Context, ...) ([]Entitlement, error) {
	entitlements := []Entitlement{}

	query := "SELECT definition, id, organizationID FROM entitlements WHERE organizationID=@organizationID"
	stmt := spanner.NewStatement(query)
	stmt.Params["organizationID"] = "1234"

	iter := ds.Single().Query(ctx, stmt)
	defer iter.Stop()

	for {
		row, err := iter.Next()
		if err == iterator.Done {
			break
		}
		if err != nil {
                         // GETTING ERROR HERE
			return nil, err
		}

		e := Entitlement{}
                err = row.ToStruct(&e)
                if err != nil {
                        return nil, err
                }

                entitlements = append(entitlements, e)
	}

	return entitlements, nil
}
```

After some load to the API that fronts this DB logic, I start getting errors from Spanner that look like this:

```
'spanner: code = "Canceled", desc = "timeout / context canceled during getting session"'
```

It seems like there is some kind of Spanner Session leaking going on, and I'm not able to figure out why. Shouldn't the `defer iter.Stop()` terminate the Session that is being consumed to execute the Spanner Query? It seems like there is a leak with the underlying `ReadOnlyTransaction`.

Help would be appreciated.

**Comment by olavloite** [comment]
@jon-whit Thanks for your detailed report.

Yes, the `defer iter.Stop()` should ensure that the session is returned to the session pool and should be enough to avoid any session leaks. The example code you supplied also seems fine. I have a couple of questions, though:
* Do you use any custom session pool configuration when creating your `Spanner` client? I'm especially interested in any value you might have set for `SessionPoolConfig.MaxOpened` and/or `SessionPoolConfig.MaxBurst`.
* The context that is used to execute the query is passed in to the `ListEntitlements` method. Does that context have a deadline/timeout? And is it re-used for multiple queries?
* Is the `ListEntitlements` method called concurrently by multiple go routines?

**Comment by jon-whit** [comment]
@olavloite thanks for following up.

The overall code looks something like this:

`main.go`
```go
package main

ctx := context.Background()
client, err := spanner.NewClientWithConfig(ctx, databaseName, spanner.ClientConfig {
	SessionPoolConfig: spanner.SessionPoolConfig {
		MinOpened: cfg.SpannerMinOpenedSessions,
	},
})

spannerDatastore := datastores.SpannerDatastore {
    client,
}

controller := controllers.HTTPController {
    EntitlementsService: &services.Entitlements {
        Datastore: spannerDatastore,
        ....
}

r := mux.NewRouter()
controller.InitRoutes(r)
```

`http_controller.go`
```
package controllers

type HTTPController struct {
	EntitlementsService interfaces.EntitlementsService
}

func (h *HTTPController) InitRoutes(r *mux.Router) {
        public := r.PathPrefix("/foundation/entitlements").Subrouter()

	public.Methods(http.MethodGet).Path("")..HandlerFunc(h.listEntitlements)
}

func (h *HTTPController) listEntitlements(w http.ResponseWriter, req *http.Request) {
    entitlements, err := h.EntitlementsService.ListEntitlements(req.Context())
    if err != nil {
        // handle error
    }

    h.JSON(w, http.StatusOK, entitlements) // Write the JSON response
}
```

`entitlements_service.go`
```
package services

type Entitlements struct {
    Datastore interfaces.EntitlementsDatastore
}

func (e *Entitlements) ListEntitlements(ctx context.Context) ([]Entitlements, error) {
    e, err := e.Datastore.ListEntitlements(ctx)
    if err != nil {
        // handle error
   }

   return e, nil
}
```

`spanner_datastore.go`
```go
package datastores

// SpannerDatastore implements the EntitlementsDatastore interface
type SpannerDatastore struct {
	*spanner.Client
}

func (ds *SpannerDatastore) ListEntitlements(ctx context.Context) ([]Entitlements, error) {
    // Code is in first post.
}
```

----

So to answer your questions:

* Yes, we use the `SessionsPoolConfig` MinOpened parameter to attempt to keep a minimum number of sessions available. We noticed this helped improve our query workload performance by bumping this number slightly.

* There is no additional deadline/timeout applied to the context that is passed down from the `HTTPController` to the `SpannerDatastore`. The request that is passed along comes from the originating HTTP request Context.

* The `SpannerDatastore`'s ListEntitlements method is called from multiple Go routines, because it is invoked as a function when handling an incoming HTTP request.

Hope this info helps. Does anything obvious stand out?

**Comment by jon-whit** [comment]
The `ListEntitlements` query is the most frequently executed query, so I think you're right that the issue is surfacing itself when executing that query because it has a higher probability of doing so.

I went over the codebase this morning looking for any queries that executing a `client.Single().Query(...)` call without a corresponding `defer iter.Stop()`. I found two places in particular that there were `defer iter.Stop()` calls missing.

Would failing to call the RowIterator's Stop method cause the Session leak? If so, I think that's the culprit. I still need to test to see if that's the case.
